### PR TITLE
Add support for VNC console for shim VMs

### PR DESCRIPTION
**Pull request by europaul**
We enable access to shim VMs via VNC console through the following steps:
- add another virtual console in the QEMU configuration that's gonna be used by VNC
- call agetty on that virtual console during VMs boot to enable login

When connecting to the QEMU instance via VNC the user will first be presented with a view of the currently running process (container's entry point). To switch to the shell of the shim VM the user will have to press 'Ctrl+Alt+2' and then 'Enter' to get the login prompt. Due to inability of some VNC clients to change the resolution of the display, the client might crash. However upon the client restart the new virtual console will appear as expected. The user can then switch back to the previous virtual console by pressing 'Ctrl+Alt+1'.

### CODE REVIEW

File: pkg/xen-tools/initrd/init-initrd [hunk 1]
```diff
@@ -164,21 +164,25 @@ maxupdateskew 100.0
 EOF
     /usr/sbin/chronyd -u root -f /etc/chrony/chrony.conf &&
         echo "chronyd started!" > /dev/kmsg ||
-        echo "chronyd NOT started!" > /dev/ksmg
```
Comment (rouming) [inline pkg/xen-tools/initrd/init-initrd:167]: That's a good catch! But I beg you to make a separate commit.

**Comment by rouming** [comment]
Can you please add to the commit message and PR description that this will work only for the standalone vnc client, not the "Remote Console" guacamole plugin. Also would be great if you specify the reasons why it does not work for there.

I do not have any objections on this PR, this is what we discussed with you. The main thing is left to be done here is to enable this feature by a separate flag, like the regular vnc is enabled: https://github.com/lf-edge/eve-api/blob/6180247abbc1c04422e093458809e8da59a25eff/proto/config/vm.proto#L52

Why? There are huge security concerns that we can't let everyone who has a vnc access to the Container let freely access the Shim Vm (if you wish we can discuss that with you and @shjala once again). 

So my advice is in order to finish with this PR is to add a new field to the `VmConfig` structure (vm.proto). Then we can raise the whole security discussion how to enable this safely.

**Comment by europaul** [comment]
@rouming I added the flag `EnableVncShimVm` to pillar. It needs to be added to edgeview as well, but since it depends on pillar I will have to create another PR for that after this one is merged.

File: pkg/pillar/hypervisor/kvm.go [hunk 2]
```diff
@@ -169,6 +169,7 @@ const qemuConfTemplate = `# This file is automatically generated by domainmgr
   logfile = "/dev/fd/1"
   logappend = "on"
 
+{{- if .DomainConfig.EnableVncShimVm}}
```
Comment (eriknordmark) [inline pkg/pillar/hypervisor/kvm.go:172]: This implies that it needs to be enabled before the app instance is booted. That needs to be documented.

**Comment by eriknordmark** [comment]
> Can you please add to the commit message and PR description that this will work only for the standalone vnc client, not the "Remote Console" guacamole plugin. Also would be great if you specify the reasons why it does not work there.

I'm also curious why it wouldn't work since guacamole is built using the vnc access in kvm/qemu.

Don't we have some markdown file which documents the current use of the console/vnc where this information can be added? If not we need to add it.

File: pkg/pillar/hypervisor/xen.go [hunk 3]
```diff
@@ -198,7 +198,7 @@ func (ctx xenContext) CreateDomConfig(domainName string, config types.DomainConf
 		file.WriteString(fmt.Sprintf("bootloader = \"%s\"\n",
 			bootLoader))
 	}
-	if config.EnableVnc {
+	if config.EnableVnc || config.EnableVncShimVm {
```
Comment (christoph-zededa) [inline pkg/pillar/hypervisor/xen.go:201]: I am not sure that it is expected from the user that if `EnableVncShimVm` is set, that also normal VNC is enabled.

File: pkg/pillar/hypervisor/kvm.go (hunk 2 above)
Comment (rouming) [reply pkg/pillar/hypervisor/kvm.go:172]: Documentation is a very good point. You also hide the prime console chardev under this "if" branch, which means it won't be created on the guest side if vnc for shim is disabled. Probably you meant to hide the "[chardev "charserial2"]", line 177.

File: pkg/pillar/hypervisor/xen.go (hunk 3 above)
Comment (rouming) [reply pkg/pillar/hypervisor/xen.go:201]: That is a good suspect - either we enable vnc -> shim vnc in this order, either we enable vnc and shim vnc separately.

**Comment by rouming** [comment]
@europaul There are a few changes requested. Also please fix yetus. Seems it wants "VM" name (all capital).

**Comment by rouming** [comment]
> > Can you please add to the commit message and PR description that this will work only for the standalone vnc client, not the "Remote Console" guacamole plugin. Also would be great if you specify the reasons why it does not work there.
> 
> I'm also curious why it wouldn't work since guacamole is built using the vnc access in kvm/qemu.

Here we talk about this https://www.qemu.org/docs/master/system/keys.html, so you can switch between consoles from the VNC viewport. But guacamole browser client (I assume this is guacamole) does not let you send those keys combinations (or this is the browser? I did not invesitate that, might be Paul knows). So the switch between VNC consoles (vm and shim-vm) will work from application client (tightvnc, etc), but not guacamole. 

> 
> Don't we have some markdown file which documents the current use of the console/vnc where this information can be added? If not we need to add it.

Seems indeed we don't have any documentation VNC related.

File: pkg/pillar/hypervisor/kvm.go (hunk 2 above)
Comment (europaul) [reply pkg/pillar/hypervisor/kvm.go:172]: nice catch @rouming , thank you

File: pkg/pillar/hypervisor/xen.go (hunk 3 above)
Comment (europaul) [reply pkg/pillar/hypervisor/xen.go:201]: thank you @christoph-zededa , I'd rather have it cascading like
EnableVnc                      [ ]
|_ EnableVncShimVM    [ ]
so EnableVncShimVM cannot be set if EnableVnc is not set. The controller should be implemented accordingly - I'll mention this in the documentation.

**Comment by europaul** [comment]
@eriknordmark @europaul I'm still looking into how to switch between consoles using a key combination in guacamole. According to the guacamole's [FAQ](https://guacamole.apache.org/faq/#keyboard-shortcuts) this might be possible. If I find a solution it will be part of another PR.

File: pkg/pillar/hypervisor/xen.go (hunk 3 above)
Comment (christoph-zededa) [reply pkg/pillar/hypervisor/xen.go:201]: Perhaps it is also possible to check it in eve and send back an error to the controller?

Comment (europaul) [reply pkg/pillar/hypervisor/xen.go]: Yes, did it.

File: pkg/pillar/hypervisor/kvm.go (hunk 2 above)
Comment (europaul) [reply pkg/pillar/hypervisor/kvm.go]: Done

File: pkg/pillar/cmd/zedagent/parseconfig.go [hunk 4]
```diff
@@ -2614,6 +2615,12 @@ func checkAndPublishAppInstanceConfig(getconfigCtx *getconfigContext,
 		config.Errors = append(config.Errors, err.Error())
 	}
 
+	if config.FixedResources.EnableVnc == false && config.FixedResources.EnableVncShimVM == true {
+		err := fmt.Errorf("VNC shim VM enabled but VNC disabled for app instance %s", config.UUIDandVersion.UUID)
+		log.Error(err)
+		config.Errors = append(config.Errors, err.Error())
```
Comment (christoph-zededa) [inline pkg/pillar/cmd/zedagent/parseconfig.go:2621]: What happens if the error is set? Does it disallow starting the VM?

Comment (christoph-zededa) [reply pkg/pillar/cmd/zedagent/parseconfig.go:2621]: I see the comment
```
  102      // Error                                                                                                                                                       
  103      //  If this is set, do not process further.. Just set the status to error                                                                                      
  104      //  so the cloud gets it.  
```
but I don't see where processing it is actually stopped

Comment (christoph-zededa) [reply pkg/pillar/cmd/zedagent/parseconfig.go:2621]: Thx, now I see it!

**Comment by europaul** [comment]
@eriknordmark @rouming I think I addressed all comments, feel free to have another look!

**Comment by europaul** [comment]
@rouming @eriknordmark I actually just confirmed that switching between virtual consoles works via the keyboard shortcut in guacamole in both Firefox in Chromium on my machine. So I removed that line about the limitation from the docs.

**Comment by rouming** [comment]
That's cool. Thanks for update.

File: pkg/pillar/hypervisor/kvm.go [hunk 5]
```diff
@@ -173,6 +173,16 @@ const qemuConfTemplate = `# This file is automatically generated by domainmgr
   driver = "virtconsole"
   chardev = "charserial1"
   name = "org.lfedge.eve.console.prime"
+
+{{- if .DomainConfig.EnableVncShimVM}}
```
Comment (christoph-zededa) [inline pkg/pillar/hypervisor/kvm.go:177]: What about adding a test to check that this is actually activated like the other tests we already have for this template?

Comment (europaul) [reply pkg/pillar/hypervisor/kvm.go:177]: Added it

File: pkg/pillar/types/domainmgrtypes.go [hunk 6]
```diff
@@ -152,6 +152,7 @@ func (config DomainConfig) LogCreate(logBase *base.LogObject) {
 	}
 	logObject.CloneAndAddField("activate", config.Activate).
 		AddField("enable-vnc", config.EnableVnc).
```
[the remark this answers is not in the document]
Comment (europaul) [reply pkg/pillar/types/domainmgrtypes.go:154]: Done

File: pkg/pillar/hypervisor/kvm_test.go [hunk 7]
```diff
@@ -2319,3 +2618,36 @@ func TestCreateDom(t *testing.T) {
...
+	diskStatuses = diskStatuses[1:]
+
+	config, aa := domainConfigAndAssignableAdapters(diskConfigs)
+	// enable VNC and VNC for shim VM
+	config.VmConfig.EnableVnc = true
+	config.VmConfig.EnableVncShimVM = true
+	if err := kvmIntel.CreateDomConfig("test", config, types.DomainStatus{}, diskStatuses, &aa, conf); err != nil {
+		t.Errorf("CreateDomConfig failed %v", err)
+	}
+	defer os.Truncate(conf.Name(), 0)
```
Comment (christoph-zededa) [inline pkg/pillar/hypervisor/kvm_test.go:2643]:
```suggestion
	
```
why is it needed for if you remove the file anyways?

**Review by eriknordmark — APPROVED** [review]
LGTM (but I think there is a comment or two from @christoph-zededa to look at)
